curvefs/src/mds/fs_manager.cpp

```diff
@@ -0,0 +1,424 @@
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Created Date: 2021-05-19
+ * Author: chenwei
+ */
+#include <glog/logging.h>
+#include <sys/stat.h>  // for S_IFDIR
+#include <limits>
+#include "curvefs/src/mds/fs_manager.h"
+
+namespace curvefs {
+namespace mds {
+FSStatusCode FsManager::CreateFs(const std::string& fsName, uint64_t blockSize,
+                            const Volume& volume, FsInfo* fsInfo) {
+    // 1. query fs
+    if (fsStorage_->Exist(fsName)) {
+        LOG(WARNING) << "CreateFs fail, fs exist, fsName = " << fsName
+                  << ", blockSize = " << blockSize;
+        return FSStatusCode::FS_EXIST;
+    }
+
+    // 2. create fs
+    MdsFsInfo newFsInfo(GetNextFsId(), fsName, FsStatus::NEW, GetRootId(),
+                            volume.volumesize(), blockSize, volume);
+
+    // 3. insert fs, fs status is NEW
+    FSStatusCode ret = fsStorage_->Insert(newFsInfo);
+    if (ret != FSStatusCode::OK) {
+        LOG(ERROR) << "CreateFs fail, insert fs fail, fsName = " << fsName
+                   << ", blockSize = " << blockSize
+                   << ", ret = " << FSStatusCode_Name(ret);
+        return ret;
+    }
+
+    // 4. use metaserver interface, insert rootinode
+    uint32_t fsId = newFsInfo.GetFsId();
+    uint32_t uid = 0;  // TODO(cw123)
+    uint32_t gid = 0;  // TODO(cw123)
+    uint32_t mode = S_IFDIR | 0777;  // TODO(cw123)
+    ret = metaserverClient_->CreateRootInode(fsId, uid, gid, mode);
+    if (ret != FSStatusCode::OK) {
+        LOG(ERROR) << "CreateFs fail, insert root inode fail"
+                   << ", fsName = " << fsName
+                   << ", blockSize = " << blockSize
+                   << ", ret = " << FSStatusCode_Name(ret);
+        FSStatusCode ret2 = fsStorage_->Delete(fsName);
+        if (ret2 != FSStatusCode::OK) {
+            LOG(ERROR) << "CreateFs fail, insert root inode fail,"
+                       << " then delete fs fail, fsName = " << fsName
+                        << ", errCode = " << ret2;
+            return ret2;
+        }
+        return ret;
+    }
+
+    // 5. update fs status to INITED
+    newFsInfo.SetStatus(FsStatus::INITED);
+    ret = fsStorage_->Update(newFsInfo);
+    if (ret != FSStatusCode::OK) {
+        LOG(ERROR) << "CreateFs fail, update fs to inited fail"
+                   << ", fsName = " << fsName
+                   << ", blockSize = " << blockSize
+                   << ", ret = " << FSStatusCode_Name(ret);
+        ret = metaserverClient_->DeleteInode(fsId, GetRootId());
+        if (ret != FSStatusCode::OK) {
+            LOG(ERROR) << "CreateFs fail, update fs status to inited fail"
+                       << ", then delete root inode fail"
+                        << ", fsName = " << fsName
+                        << ", blockSize = " << blockSize
+                        << ", ret = " << FSStatusCode_Name(ret);
+            return ret;
+        }
+
+        ret = fsStorage_->Delete(fsName);
```

Comment:
if fsStorage_ Update fail, there is not a  FsInfo in storage, so Delete is not needed. But you can still call Delete Function,  and Delete Function must not return an error when FsInfo not found, otherwise the if as follows will print a error log.

Reply:
The way of implementation has changed.This piece of code is commented.